src/BotPlutusInterface/Balance.hs

```diff
@@ -74,62 +78,69 @@ import BotPlutusInterface.Collateral (removeCollateralFromMap)
 import Prettyprinter (pretty, viaShow, (<+>))
 import Prelude
 
--- `BalanceTx` is for checking which type of `Tx` we are balancing.
-data BalanceTx
-  = -- | `BalanceTxCollateral` denotes that we are balancing the `Tx` that creates collateral.
-    BalanceTxCollateral
-  | -- | `BalanceTxWithScripts` denotes that we are balancing the `Tx` that uses scripts.
-    BalanceTxWithScripts
-  | -- | `BalanceTxWithoutScripts` denotes that we are balancing the `Tx` that doesn't uses any scripts.
-    BalanceTxWithoutScripts
-  deriving stock (Eq)
-
--- `balanceTxIO` checks the type of the current `UnbalancedTx` and then calls `balanceTxIO'`.
+data TxWithScript
+data TxWithoutScript
+
+data BalanceTxConstraint a where
+  TxWithScript :: BalanceTxConstraint TxWithScript
+  TxWithoutScript :: BalanceTxConstraint TxWithoutScript
+  TxWithSeparateChange :: BalanceTxConstraint a
+
+instance Eq (BalanceTxConstraint a) where
+  TxWithScript == TxWithScript = True
+  TxWithoutScript == TxWithoutScript = True
+  TxWithSeparateChange == TxWithSeparateChange = True
+  _ == _ = False
+
+class KnownBalanceConstraint (a :: Type) where
+  knownConstraint :: proxy a -> BalanceTxConstraint a
```

Comment:
a kind sig on proxy would be nice, also - why are we using proxies here? we have type application.

Reply:
Fixed, The proxy was redundant. Thanks for pointing it out.